src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+}
+
+bool Analyser::AnalyserImpl::isStateVariable(const AnalyserInternalVariablePtr &variable)
+{
+    return variable->mType == AnalyserInternalVariable::Type::STATE;
+}
+
+bool Analyser::AnalyserImpl::isConstantOrAlgebraicVariable(const AnalyserInternalVariablePtr &variable)
+{
+    return (variable->mType == AnalyserInternalVariable::Type::CONSTANT)
+           || (variable->mType == AnalyserInternalVariable::Type::COMPUTED_TRUE_CONSTANT)
+           || (variable->mType == AnalyserInternalVariable::Type::COMPUTED_VARIABLE_BASED_CONSTANT)
+           || (variable->mType == AnalyserInternalVariable::Type::ALGEBRAIC);
+}
+
+bool Analyser::AnalyserImpl::compareVariablesByTypeAndIndex(const AnalyserInternalVariablePtr &variable1,
+                                                            const AnalyserInternalVariablePtr &variable2)
+{
+    if (isStateVariable(variable1) && isConstantOrAlgebraicVariable(variable2)) {
+        return true;
+    }
+
+    if (isConstantOrAlgebraicVariable(variable1) && isStateVariable(variable2)) {
+        return false;
+    }
+
+    return variable1->mIndex < variable2->mIndex;
+}
+
+bool Analyser::AnalyserImpl::compareEquationsByVariable(const AnalyserInternalEquationPtr &equation1,
+                                                        const AnalyserInternalEquationPtr &equation2)
+{
+    return compareVariablesByTypeAndIndex(equation1->mVariable, equation2->mVariable);
+}
+
+size_t Analyser::AnalyserImpl::mathmlChildCount(const XmlNodePtr &node) const
+{
+    // Return the number of child elements, in the MathML namespace, for the
+    // given node.
+
+    auto childNode = node->firstChild();
+    size_t res = childNode->isMathmlElement() ? 1 : 0;
+
+    while (childNode != nullptr) {
+        childNode = childNode->next();
+
+        if (childNode && childNode->isMathmlElement()) {
+            ++res;
+        }
+    }
+
+    return res;
+}
+
+XmlNodePtr Analyser::AnalyserImpl::mathmlChildNode(const XmlNodePtr &node,
+                                                   size_t index) const
+{
+    // Return the nth child element of the given node, skipping anything that is
+    // not int he MathML namespace.
+
+    auto res = node->firstChild();
+    auto childNodeIndex = res->isMathmlElement() ? 0 : MAX_SIZE_T;
+
+    while ((res != nullptr) && (childNodeIndex != index)) {
+        res = res->next();
+
+        if (res && res->isMathmlElement()) {
+            ++childNodeIndex;
+        }
+    }
+
+    return res;
+}
+
+AnalyserInternalVariablePtr Analyser::AnalyserImpl::analyserVariable(const VariablePtr &variable)
+{
+    // Find and return, if there is one, the analyser variable associated with
+    // the given variable.
+
+    for (const auto &internalVariable : mInternalVariables) {
+        if (isSameOrEquivalentVariable(variable, internalVariable->mVariable)) {
+            return internalVariable;
+        }
+    }
+
+    // No analyser variable exists for the given variable, so create one, track
```

Comment:
This behaves differently to every other fetching function we have? is that ok?
ie: normally asking for `model->component("doesnt_exist")` doesn't create that component?

Reply:
You are comparing API methods with private implementation methods. For me, it's like comparing apples and oranges. If `Analyser::AnalyserImpl::analyserVariable()` was only to return an existing (analyser) variable or `nullptr` then it means that wherever I call that method, I would need to check what was returned an create an (analyser) variable if I got `nullptr`. This would make the code unnecessarily complex, hence the current (private) implementation.